website/content/blog/Towards-Optimized-Microservices-Performance-Sustainability-via-Istio-Kepler-and-Smart-Scheduling.md

```diff
@@ -0,0 +1,64 @@
+
+![Kepler Architecture](images/241-kepler.png)
+
+For bare-metal environments, Kepler employs CPU performance counters, monitoring aspects like CPU cycles, instructions, and cache misses. It also utilizes RAPL to provide energy readings. Furthermore, it employs a usage-based ratio method, attributing energy consumption to processes based on the percentage of total CPU instructions consumed.
+
+In virtual machine settings where RAPL access is unavailable, Kepler adopts a machine learning prediction approach. It employs ML models along with BPF context switch events and stats to predict energy consumption at the container level. These techniques form the core of Kepler's methodology, ensuring precise and efficient power consumption estimations across different infrastructures.
+
+## Smart Scheduling with AI
+
+In the current experimental setup, we demonstrated a Kubernetes cluster with three nodes: one master node and two worker nodes. All nodes were equipped with Kepler and Istio deployments. The master node additionally hosted the load generation application and smart scheduling logic. Within the worker nodes, data storage services and business logic services were deployed to different nodes based on varying scheduling policies. 
+
+Regarding the testing environment, all three nodes had identical configurations: 8 cores and 32 GB of RAM. The software stack was based on Ubuntu 22.04 and included Kubernetes, DBS, Nginx, Istio, Kepler, and Wrk. We conducted tests with different workloads, ranging from 2 to 128 threads and corresponding increments in connection numbers (multiplied by 10). Input data varied from 2000 to 10,000. Each test lasted for 60 seconds.
+
+Scenario 1: Default Scheduling Policy
+
+In the first scenario, we employ Kubernetes' default scheduling policy, allowing the platform's innate intelligence to allocate resources. The "Business Logic" Service and the "Data Store" Service are distributed across Worker01 and Worker02 based on Kubernetes' default algorithm. This scenario relies on Kubernetes' understanding of workload demands, node capacities, and resource availability. The approach embodies simplicity, leveraging Kubernetes' built-in logic to ensure a balanced distribution of services. However, while default scheduling is convenient, it may not always cater to specific workload nuances, potentially leading to suboptimal resource utilization and performance imbalances.
+
+Scenario 2: Customized Service Allocation
+
+In contrast, the second scenario adopts a more targeted approach. Here, the "Data Store" Service is strategically placed on Worker01, while the "Business Logic" Service finds its home on Worker02. This customization allows for meticulous resource allocation, aligning services with nodes that best suit their computational and memory needs.
+
+![Typical Scheduling Policy](images/241-scheduling.png)
```

Comment:
```suggestion
<p class="mt-5 mb-5"><img src="/images/blogs/2023-09-cloud-native-sustainability-week/241-scheduling.png"></p>
```

@jiangphcn could you move the image to `/images/blogs/2023-09-cloud-native-sustainability-week/241-scheduling.png`